common/k8s/src/middleware/metadata.rs

```diff
@@ -1,281 +1,204 @@
+                let obj_ref = ObjectRef::from_obj(&pod);
+                if reader.get(&obj_ref).is_none() {
+                    Metrics::k8s().increment_creates();
+                }
+            }
+            WatcherEvent::Deleted(_) => {
                 Metrics::k8s().increment_deletes();
-            } // remove
+            }
             WatcherEvent::Restarted(pods) => {
-                let mut metadata = self.metadata.lock();
-                for pod in pods {
-                    let pod_meta_data = PodMetadata::try_from(pod)?;
-                    metadata
-                        .entry((pod_meta_data.name.clone(), pod_meta_data.namespace.clone()))
-                        .and_modify(|e| {
-                            Metrics::k8s().increment_creates();
-                            *e = pod_meta_data.clone()
-                        })
-                        .or_insert_with(|| {
-                            Metrics::k8s().increment_creates();
-                            pod_meta_data
-                        });
+                for _ in pods {
+                    Metrics::k8s().increment_creates();
                 }
             }
         }
         Ok(())
     }
-
-    async fn add_delay(&self, backoff: &mut ExponentialBackoff) {
-        let mut interval = backoff.next_backoff();
-        if interval.is_none() {
-            interval = Some(backoff.max_interval);
-        }
-        if let Some(duration) = interval {
-            tokio::time::sleep(duration).await;
-        }
-    }
 }
 
 impl Middleware for K8sMetadata {
-    fn run(&self) {
-        // Start parsing k8s events in the background
-        let runtime = Builder::new_multi_thread().build().unwrap();
-        runtime.block_on(async move {
-            let backoff = Rc::new(RefCell::new(ExponentialBackoff::default()));
-            let watcher = watcher(self.api.clone(), ListParams::default());
-
-            watcher
-                .into_stream()
-                .filter_map(|r| async {
-                    let mut backoff = backoff.borrow_mut();
-                    match r {
-                        Ok(event) => {
-                            backoff.reset();
-                            Some(event)
-                        }
-                        Err(e) => {
-                            log::warn!("k8s watch stream error: {}", e);
-                            // When polled after a some errors, the watcher will try to recover.
-                            // We should avoid eagerly polling in those cases.
-                            self.add_delay(&mut backoff).await;
-                            None
-                        }
-                    }
-                })
-                .for_each(|p| async {
-                    self.handle_pod(p)
-                        .unwrap_or_else(|e| log::warn!("unable to process pod event: {}", e));
-                })
-                .await;
-        });
-    }
+    fn run(&self) {}
```

Comment:
I assume the reason the whole run function is able to go away is because the reflector is handling all the data management internally?

Reply:
Yeh, it's just here to fulfill the old synchronous Middleware interface. The driver returned from new will "run" it, so this doesn't need to do anything, but it needs to exist to avoid having to refactor `Middleware`.